### PR TITLE
use sync.Pool for validateState instances

### PR DESCRIPTION
**Pull request by rogpeppe**
This speeds things up by reducing allocations.

```
name        old time/op  new time/op  delta
Validate-4   547ns ± 1%   418ns ± 3%  -23.56%  (p=0.008 n=5+5)
```

Doing a basic validation now does zero allocations in the usual
case.

### CODE REVIEW

File: validator.go [hunk 1]
```diff
@@ -233,12 +233,11 @@ func (mv *Validator) Validate(x interface{}) error {
...
 	// we could keep validateState instances around in a sync.Pool
 	// to avoid the allocations.
-	state := &validateState{
-		path:      make([]byte, 0, 20),
-		pathStack: make([]int, 0, 10),
-	}
+	state := getState()
 	validate(sv, state)
-	return state.finalError()
+	err := state.finalError()
```
Comment (aranw) [inline validator.go:238]: Is there a reason why this happens before `putState`?

File: validator.go [hunk 2]
```diff
@@ -281,12 +280,11 @@ func (mv *Validator) Valid(v interface{}, tag string) error {
...
 		return err
 	}
-	state := &validateState{
-		path:      make([]byte, 0, 20),
-		pathStack: make([]int, 0, 10),
-	}
+	state := getState()
 	validate(sv, state)
-	if err := state.finalError(); err != nil {
+	err = state.finalError()
```
Comment (aranw) [inline validator.go:285]: Same as above. Is there a reason for this to happen before the `putState` call?

File: validator.go (hunk 1 above)
Comment (tealeg) [reply validator.go:238]: @aranw I'm guessing it's because `putState` clears the errors to nil, so you'd lose them if you ordered it differently.

**Review by tealeg — APPROVED** [review]
:+1:

File: validator.go (hunk 2 above)
Comment (rogpeppe) [reply validator.go:285]: as @tealeg says.

**Pull request merged**